manifests/agent/windows.pp

```diff
@@ -1,73 +1,71 @@
-class distelli::agent::windows (
-  $version = '3.66.33',
-  $tempdir = 'C:/Users/distelli/AppData/Local/Temp/1',
-){
+# Install the Puppet Pipelines agent on Windows
+#
+# Do not use directly; use pipelines::agent.
+class pipelines::agent::windows {
+  $env = $facts['pipelines_env']
 
-  if $::facts['os']['hardware'] == 'x86_64' {
-    $archive = "distelli.Windows-AMD64-${version}.gz"
-    $url = "https://s3.amazonaws.com/download.distelli.com/distelli.Windows-AMD64/${archive}"
-  }
-  else {
-    $archive = "distelli.Windows-x86-${version}.gz"
-    $url = "https://s3.amazonaws.com/download.distelli.com/distelli.Windows-x86/${archive}"
-  }
-
-  $workdir = 'C:/Program Files/Distelli'
-
-  archive { "${tempdir}/${archive}" :
-    source       => $url,
-    # cleanup      => false,
-    creates      => "${tempdir}/distelli",
-    extract      => true,
-    extract_path => $tempdir,
-    require      => User['distelli'],
+  if ! $pipelines::agent::install_dir {
+    $install_dir = "${env['ProgramFiles']}\\Distelli"
+  } else {
+    $install_dir = $pipelines::agent::install_dir
   }
 
-  file { $workdir :
-    ensure  => directory,
-    owner   => 'distelli',
-    group   => 'Administrators',
-    require => User['distelli'],
+  if $pipelines::agent::version {
+    $download_url = "${pipelines::agent::download_url}/${pipelines::agent::version}.ps1"
+  } else {
+    $download_url = "${pipelines::agent::download_url}.ps1"
   }
+  $download_location = "${install_dir}\\distelli-download.ps1"
+  $download_cmd = "cmd.exe /c \"powershell -NoProfile -ExecutionPolicy Bypass -Command - < \"${download_location}\"\""
+  $agent_conf_file = "${env['SystemDrive']}\\distelli.yml"
 
-  # Requires fqdn_rand_string function from puppetlabs/stdlib
-  $distelli_exec = "distelli-${fqdn_rand_string(5)}.exe"
-
-  exec { 'Copy executable' :
-    command   => "Copy-Item ${tempdir}/distelli \$ENV:ProgramFiles/Distelli/${distelli_exec}",
-    unless    => "If (Test-Path -Path \$ENV:ProgramFiles/Distelli/${distelli_exec}) { exit 0 } else { exit 1}",
-    provider  => powershell,
-    require   => File[$workdir],
-    logoutput => true,
+  exec { "mkdir ${install_dir}":
+    command => "cmd.exe /c \"md \"${install_dir}\"\"",
+    path    => $facts['path'],
+    onlyif  => "cmd.exe /c \"if exist \"${install_dir}\" exit 1\"",
   }
-
-  file { ["${workdir}/distelli.exe",  "${workdir}/dagent.exe", "${workdir}/dtk.exe"] :
-    ensure  => link,
-    target  => "${workdir}/${distelli_exec}",
-    require => Exec['Copy executable'],
+  file { $download_location:
+    source  => $download_url,
+    require => Exec["mkdir ${install_dir}"],
   }
-
-  exec { 'Test distelli.exe execution' :
-    command     => "& \$ENV:ProgramFiles/Distelli/distelli.exe version",
+  exec { 'pipelines::agent download':
     provider    => powershell,
```

Comment:
I would remove the dependency on the PowerShell module here if you can help it. The fewer dependencies the better, and it doesn't look like you have a compelling need for the module in this case.  (You can still call out to `powershell.exe` yourself using the normal `exec` as you've done)

Reply:
I see that @RandomNoun7 marked this as resolved, was this intentional? FWIW, I originally tried to not use the powershell module (as you might have noticed in the unused $download_cmd), but it just didn't work, but maybe there was something else that was causing problems and I can revert back to not using the powershell module? It seems that recommendation in the Windows hipchat room was to use powershell? Guidance is appreciated :).

Reply:
I don't think I intended to mark this as resolved. Maybe I hit a button with my thumb while browsing these PR comments on my phone.

Reply:
Okay, here is what happens if I use $download_cmd and NOT the powershell provider:

```
Debug: Exec[pipelines::agent download](provider=windows): Executing 'cmd.exe /c "powershell -NoProfile -ExecutionPolicy Bypass -Command - < "C:\Program Files\Distelli\distelli-download.ps1""'
Debug: Executing: 'cmd.exe /c "powershell -NoProfile -ExecutionPolicy Bypass -Command - < "C:\Program Files\Distelli\distelli-download.ps1""'
Notice: /Stage[main]/Pipelines::Agent::Windows/Exec[pipelines::agent download]/returns:
Notice: /Stage[main]/Pipelines::Agent::Windows/Exec[pipelines::agent download]/returns:
Notice: /Stage[main]/Pipelines::Agent::Windows/Exec[pipelines::agent download]/returns:     Directory: C:\cygwin64\tmp
Notice: /Stage[main]/Pipelines::Agent::Windows/Exec[pipelines::agent download]/returns:
Notice: /Stage[main]/Pipelines::Agent::Windows/Exec[pipelines::agent download]/returns:
Notice: /Stage[main]/Pipelines::Agent::Windows/Exec[pipelines::agent download]/returns: Mode                LastWriteTime     Length Name
Notice: /Stage[main]/Pipelines::Agent::Windows/Exec[pipelines::agent download]/returns: ----                -------------     ------ ----
Notice: /Stage[main]/Pipelines::Agent::Windows/Exec[pipelines::agent download]/returns: d----        10/16/2018   4:03 PM            distelli-install
Notice: /Stage[main]/Pipelines::Agent::Windows/Exec[pipelines::agent download]/returns:
Notice: /Stage[main]/Pipelines::Agent::Windows/Exec[pipelines::agent download]/returns:
Notice: /Stage[main]/Pipelines::Agent::Windows/Exec[pipelines::agent download]/returns:     Installing Distelli Toolkit 3.67.5 (Windows-AMD64)...
Notice: /Stage[main]/Pipelines::Agent::Windows/Exec[pipelines::agent download]/returns: Downloading https://s3.amazonaws.com/download.distelli.com/distelli.Windows-AMD
Notice: /Stage[main]/Pipelines::Agent::Windows/Exec[pipelines::agent download]/returns: 64/distelli.Windows-AMD64-3.67.5.gz to C:\cygwin64\tmp\distelli-install\distell
Notice: /Stage[main]/Pipelines::Agent::Windows/Exec[pipelines::agent download]/returns: i.exe.gz
Notice: /Stage[main]/Pipelines::Agent::Windows/Exec[pipelines::agent download]/returns: Downloading https://s3.amazonaws.com/download.distelli.com/7za.exe to C:\cygwin
Notice: /Stage[main]/Pipelines::Agent::Windows/Exec[pipelines::agent download]/returns: 64\tmp\distelli-install\7za.exe
Notice: /Stage[main]/Pipelines::Agent::Windows/Exec[pipelines::agent download]/returns:
Notice: /Stage[main]/Pipelines::Agent::Windows/Exec[pipelines::agent download]/returns: 7-Zip (A) 9.20  Copyright (c) 1999-2010 Igor Pavlov  2010-11-18
Notice: /Stage[main]/Pipelines::Agent::Windows/Exec[pipelines::agent download]/returns:
Notice: /Stage[main]/Pipelines::Agent::Windows/Exec[pipelines::agent download]/returns: Processing archive: C:\cygwin64\tmp\distelli-install\distelli.exe.gz
Notice: /Stage[main]/Pipelines::Agent::Windows/Exec[pipelines::agent download]/returns:
Notice: /Stage[main]/Pipelines::Agent::Windows/Exec[pipelines::agent download]/returns: Extracting  distelli
Notice: /Stage[main]/Pipelines::Agent::Windows/Exec[pipelines::agent download]/returns:
Notice: /Stage[main]/Pipelines::Agent::Windows/Exec[pipelines::agent download]/returns: Everything is Ok
Notice: /Stage[main]/Pipelines::Agent::Windows/Exec[pipelines::agent download]/returns:
Notice: /Stage[main]/Pipelines::Agent::Windows/Exec[pipelines::agent download]/returns: Size:       4245641
Notice: /Stage[main]/Pipelines::Agent::Windows/Exec[pipelines::agent download]/returns: Compressed: 2321276
Notice: /Stage[main]/Pipelines::Agent::Windows/Exec[pipelines::agent download]/returns: out-lineoutput : The OS handle's position is not what FileStream expected. Do n
Notice: /Stage[main]/Pipelines::Agent::Windows/Exec[pipelines::agent download]/returns: ot use a handle simultaneously in one FileStream and in Win32 code or another F
Notice: /Stage[main]/Pipelines::Agent::Windows/Exec[pipelines::agent download]/returns: ileStream. This may cause data loss.
Notice: /Stage[main]/Pipelines::Agent::Windows/Exec[pipelines::agent download]/returns:     + CategoryInfo          : NotSpecified: (:) [out-lineoutput], IOException
Notice: /Stage[main]/Pipelines::Agent::Windows/Exec[pipelines::agent download]/returns:     + FullyQualifiedErrorId : System.IO.IOException,Microsoft.PowerShell.Comma
Notice: /Stage[main]/Pipelines::Agent::Windows/Exec[pipelines::agent download]/returns:    nds.OutLineOutputCommand
Notice: /Stage[main]/Pipelines::Agent::Windows/Exec[pipelines::agent download]/returns:
Notice: /Stage[main]/Pipelines::Agent::Windows/Exec[pipelines::agent download]/returns: Exception calling "Start" with "1" argument(s): "The system cannot find the fil
Notice: /Stage[main]/Pipelines::Agent::Windows/Exec[pipelines::agent download]/returns: e specified"
Notice: /Stage[main]/Pipelines::Agent::Windows/Exec[pipelines::agent download]/returns: At line:1 char:50
Notice: /Stage[main]/Pipelines::Agent::Windows/Exec[pipelines::agent download]/returns: +    $process = [System.Diagnostics.Process]::Start <<<< ($startInfo)
Notice: /Stage[main]/Pipelines::Agent::Windows/Exec[pipelines::agent download]/returns:     + CategoryInfo          : NotSpecified: (:) [], ParentContainsErrorRecordE
Notice: /Stage[main]/Pipelines::Agent::Windows/Exec[pipelines::agent download]/returns:    xception
Notice: /Stage[main]/Pipelines::Agent::Windows/Exec[pipelines::agent download]/returns:     + FullyQualifiedErrorId : DotNetMethodException
Notice: /Stage[main]/Pipelines::Agent::Windows/Exec[pipelines::agent download]/returns:
Notice: /Stage[main]/Pipelines::Agent::Windows/Exec[pipelines::agent download]/returns: out-lineoutput : The OS handle's position is not what FileStream expected. Do n
Notice: /Stage[main]/Pipelines::Agent::Windows/Exec[pipelines::agent download]/returns: ot use a handle simultaneously in one FileStream and in Win32 code or another F
Notice: /Stage[main]/Pipelines::Agent::Windows/Exec[pipelines::agent download]/returns: ileStream. This may cause data loss.
Notice: /Stage[main]/Pipelines::Agent::Windows/Exec[pipelines::agent download]/returns:     + CategoryInfo          : NotSpecified: (:) [out-lineoutput], IOException
Notice: /Stage[main]/Pipelines::Agent::Windows/Exec[pipelines::agent download]/returns:     + FullyQualifiedErrorId : System.IO.IOException,Microsoft.PowerShell.Comma
Notice: /Stage[main]/Pipelines::Agent::Windows/Exec[pipelines::agent download]/returns:    nds.OutLineOutputCommand
Notice: /Stage[main]/Pipelines::Agent::Windows/Exec[pipelines::agent download]/returns:
Notice: /Stage[main]/Pipelines::Agent::Windows/Exec[pipelines::agent download]/returns: out-lineoutput : The OS handle's position is not what FileStream expected. Do n
Notice: /Stage[main]/Pipelines::Agent::Windows/Exec[pipelines::agent download]/returns: ot use a handle simultaneously in one FileStream and in Win32 code or another F
Notice: /Stage[main]/Pipelines::Agent::Windows/Exec[pipelines::agent download]/returns: ileStream. This may cause data loss.
Notice: /Stage[main]/Pipelines::Agent::Windows/Exec[pipelines::agent download]/returns:     + CategoryInfo          : NotSpecified: (:) [out-lineoutput], IOException
Notice: /Stage[main]/Pipelines::Agent::Windows/Exec[pipelines::agent download]/returns:     + FullyQualifiedErrorId : System.IO.IOException,Microsoft.PowerShell.Comma
Notice: /Stage[main]/Pipelines::Agent::Windows/Exec[pipelines::agent download]/returns:    nds.OutLineOutputCommand
Notice: /Stage[main]/Pipelines::Agent::Windows/Exec[pipelines::agent download]/returns:
Error: 'cmd.exe /c "powershell -NoProfile -ExecutionPolicy Bypass -Command - < "C:\Program Files\Distelli\distelli-download.ps1""' returned 1 instead of one of [0]
Error: /Stage[main]/Pipelines::Agent::Windows/Exec[pipelines::agent download]/returns: change from 'notrun' to ['0'] failed: 'cmd.exe /c "powershell -NoProfile -ExecutionPolicy Bypass -Command - < "C:\Program Files\Distelli\distelli-download.ps1""' returned 1 instead of one of [0]
Notice: /Stage[main]/Pipelines::Agent::Windows/Exec[pipelines::agent install]: Dependency Exec[pipelines::agent download] has failures: true
```

Perhaps I'm not using the correct way of invoking powershell? Any advice is appreciated as I would also like not depend on the powershell puppet module if I can avoid it.

Reply:
After distracting @Iristyle I made several attempts to get the 'pipelines::agent download' exec to NOT use powershell. Note that the powershell script that is downloaded is here (in case anyone wants to review it and give suggestions):

https://pipelines.puppet.com/download/client.ps1

Pretty much everyone on the Windows HipChat recommends that this powershell script be changed so it downloads a zip file containing the distelli.exe and avoid needing to use 7za.exe to extract the gzip'ed distelli.exe. However, this change has cascading effects such as:

 [a] Existing customers are storing distelli.exe.gz with older versions. We could deal with this problem by writing a "backfill", but this is more work and one more moving piece.

 [b] Newer pipelines products are re-packaging the ZIP file appended to all executables such that the endpoints embedded within the executable point to the on-premise installation. This could would need to be modified so it can serve a ZIP'ed executable as an alternative to the current gzip'ed format. We could deal with this by writing more code.

 [c] Our existing build process needs to be modified. This is actually pretty easy, certainly much easier than [a] and [b].

Another alternative is to simply use 7za as we are today, but use the -o option and a file rename (since 7za thinks the file should be named "distelli" not "distelli.exe"). This is a pretty simple change since we are already doing a file rename. However, most expect that a gzip stream doesn't have a "filename"... and I suspect that our re-packaging mentioned in [b] above might not preserve this "filename". Therefore I'm "afraid" to rely on this embedded filename within the gzip stream.

Reply:
Here are the various permutation of attempts I made to make it so the powershell puppet module is not needed. This is the new exec line:

```
  exec { 'pipelines::agent download':
#    provider    => powershell,
    require     => File[$download_location],
    path        => $facts['path'],
    environment => [
      "DISTELLI_INSTALL_DIR=${install_dir}",
    ],
    command     => $download_cmd,
#    command     => "& \"${download_location}\"; Exit 0",
  }
```

...and here are the permutation I tried:

 [a] $download_cmd = "powershell.exe -NoProfile -NonInteractive -NoLogo -ExecutionPolicy Bypass -File \"${download_location}\"".

On my windows 7 test box, this command exits successfully, but it is obvious that nothing was actually executed.

 [b] $download_cmd = "powershell -NoProfile -NonInteractive -NoLogo -ExecutionPolicy Bypass -Command - < \"${download_location}\""

 powershell errors indicating that no other parameters can be specified after the -Command (aka the "<" is just another argument to powershell and does not act as a file redirect).

 [c]  $download_cmd = "cmd.exe /c \"powershell -NoProfile -NonInteractive -NoLogo -ExecutionPolicy Bypass -Command - < \"${download_location}\"\""

 The bazaar error "out-lineoutput : The OS handle's position is not what FileStream expected." as seen above.

Note that in all these permutations I made the suggested change of using a simple file rename rather than the more complex read/write loop which is used to redirect the StandardOutput of the 7za command.